### ISSUE TITLE
Allow multiple StreamWatchers per stream / share them

### ISSUE THREAD

**Issue by kelunik** [concurrent-php/ext-async#63]
I know that this is caused by `libuv`. `ext-async` should probably share one watcher per stream and use a factory instead of letting the user create these objects. That way the same watcher object can automatically be returned based on the stream ID.

Currently it's not possible to separate concerns for reading and writing in `amphp/byte-stream`, because the `ResourceInputStream` and `ResourceOutputStream` would each create a watcher for the same stream in case of a socket.

**Comment by martinschroeder** [comment]
It could be nice to have a factory if stream watchers are used a lot for these kind of things. The thing is i am planning to ship the stream API (#48) with the extension and provide streams for sockets and IO pipes backed by libuv in the future (should be more efficient than polling sockets and dealing with PHP resource streams, especially on Windows systems). This would drastically reduce the amount of use cases for stream watchers....

That being said a factory would still be better because it protects from errors. I will look into it sometime next week.

**Comment by kelunik** [comment]
The `epoll` happens in C either way, whether the read happens in userland or C probably doesn't matter too much. I tried to write a `StreamWatcherRegistry` that creates a `StreamWatcher` for each stream and refcounts them. Unfortunately, that means I can't really use `StreamWatcher::close()`, because they might be shared. It's probably better waiting on the build in stream API then.

**Comment by martinschroeder** [comment]
I added stream implementations for process pipes in #65 and for TCP streams in #70. I still need to address TLS support (for TCP streams), UDP sockets and STDIO pipes of the main process, that will take some time but should reduce the number of use cases for `StreamWatcher` quite a bit (it could even be remove someday as suggested in #69).

**Comment by martinschroeder** [comment]
With PR #104 merged i will be closing this issue because there should not be many use cases for stream watchers left. I might drop `StreamWatcher` after support for UNIX domain sockets / named pipes is done.
